### PR TITLE
[Test] Define `broken`/`skip` keyword argument for `@test`

### PR DESCRIPTION
**Pull request by giordano**
They make it less tedious having tests failing consistently only under specific
conditions.

### CODE REVIEW

**Comment by giordano** [comment]
Side note, I feel like `@test_broken` and `@test_skip` should have been
```julia
@test ... broken=true
@test ... skip=true
```
but that ship has sailed already.  Maybe good for v2.0?

**Comment by oscardssmith** [comment]
If we want that for 2.0, we should add the keyword version now so the ecosystem has plenty of time to transition.

**Comment by giordano** [comment]
If there is consensus about adding `broken`/`skip` keywords to `@test`, I can look into that (not very soon though)

File: stdlib/Test/src/Test.jl [hunk 1]
```diff
@@ -406,6 +408,36 @@ macro test_broken(ex, kws...)
...
+
+Run the test with [`@test_broken`](@ref) if `cond` is `true`, otherwise with
+[`@test`](@ref).  This is useful when the test is expected to pass only under
+specific conditions.
+
+# Examples
+```jldoctest
+julia> @test_broken_if true 2 + 2 == 5
+Test Broken
+  Brokenped: 2 + 2 == 5
```
Comment (giordano) [inline stdlib/Test/src/Test.jl:422]: I have absolutely no idea where this "Brokenped" was coming from :flushed:

**Comment by giordano** [comment]
I've updated the pull request to define the `broken`/`skip` keyword arguments for `@test`.  I'll later update replace some occurrences of `@test_broken`/`@test_skip` in the test suite where the new form makes more sense

File: test/complex.jl [hunk 2]
```diff
@@ -1088,16 +1088,8 @@ end
...
-            @test_broken isequal(complex(one(T)) / complex(T(-Inf), T(Inf)), complex(-zero(T), -zero(T)))
-        elseif T == Float32
-            @test isequal(complex(one(T)) / complex(T(Inf), T(-Inf)), complex(zero(T), zero(T)))
-            @test_broken isequal(complex(one(T)) / complex(T(-Inf), T(Inf)), complex(-zero(T), -zero(T)))
-        else
-            @test isequal(complex(one(T)) / complex(T(Inf), T(-Inf)), complex(zero(T), zero(T)))
-            @test isequal(complex(one(T)) / complex(T(-Inf), T(Inf)), complex(-zero(T), -zero(T)))
-        end
+        @test isequal(complex(one(T)) / complex(T(Inf), T(-Inf)), complex(zero(T), zero(T))) broken=(T==Float64)
+        @test isequal(complex(one(T)) / complex(T(-Inf), T(Inf)), complex(-zero(T), -zero(T))) broken=(T in (Float32, Float64))
```
Comment (giordano) [inline test/complex.jl:1092]: This is the kind of simplification that the new form enables

**Comment by giordano** [comment]
Bump :slightly_smiling_face:

File: stdlib/Test/src/Test.jl [hunk 3]
```diff
@@ -368,12 +370,73 @@ Test Passed
...
+    return quote
+        if $(length(broken) > 0 && esc(broken[1]))
+            # code to call do_test with execution result and original expr
+            do_broken_test($result, $orig_ex)
+        elseif $(length(skip) > 0 && esc(skip[1]))
+            record(get_testset(), $testres)
+        else
+            do_test($result, $orig_ex)
+        end
+    end
```
Comment (vtjnash) [inline stdlib/Test/src/Test.jl:439]:
```suggestion
    testres = :(do_test($result, $orig_ex))
    if length(skip) > 0
        testres = :($(esc(skip[1])) ? Broken(:skipped, $orig_ex) : $testres)
    end
    if length(broken) > 0
        testres = :($(esc(broken[1])) ? do_broken_test($result, $orig_ex) : $testres)
    end
    return testres
```

**Comment by miguelraz** [comment]
Triage asks to add a NEWS.md item to show the `broken=...` (the tests under `test/complex.jl` would be nice).

File: stdlib/Test/src/Test.jl [hunk 4]
```diff
@@ -368,12 +370,73 @@ Test Passed
...
+    testres = :(Broken(:skipped, $orig_ex))
+
+    return quote
+        if $(length(broken) > 0 && esc(broken[1]))
+            # code to call do_test with execution result and original expr
+            do_broken_test($result, $orig_ex)
+        elseif $(length(skip) > 0 && esc(skip[1]))
+            record(get_testset(), $testres)
+        else
+            do_test($result, $orig_ex)
```
Comment (JeffBezanson) [inline stdlib/Test/src/Test.jl:437]: Should avoid putting multiple copies of `result` in the output when possible.

**Comment by JeffBezanson** [comment]
Very nice feature!!

File: stdlib/Test/src/Test.jl (hunk 3 above)
Comment (staticfloat) [reply stdlib/Test/src/Test.jl:439]: I think we want to avoid calling `do_test()` at all if `length(skip) > 0`.

Comment (giordano) [reply stdlib/Test/src/Test.jl:439]: How about
```suggestion
    return quote
        if $(length(skip) > 0 && esc(skip[1]))
            record(get_testset(), Broken(:skipped, $orig_ex))
        else
            _do = $(length(broken) > 0 && esc(broken[1])) ? do_broken_test : do_test
            _do($result, $orig_ex)
        end
    end
```
which has a single copy of `result` as suggested by Jeff?

**Comment by giordano** [comment]
I believe I addressed all comments from triage:

* simplified the expression returned by `@test`, with a single copy of `result`
* added a `NEWS.md` entry showing a simplified test from `test/complex.jl`

**Comment by giordano** [comment]
Bump :slightly_smiling_face: The failure on macOS is in Downloads, and looks unrelated (I think)

File: test/complex.jl (hunk 2 above)
Comment (staticfloat) [reply test/complex.jl:1092]: this is amazing

**Review by staticfloat — APPROVED** [review]

**Comment by giordano** [comment]
Good to merge? :smiley:

**Pull request merged**